#6429 Roaring Booleans

yan-alex opened the pull request:
Extend Roaring Compression to support booleans as well, as opposed to only validity masks.
Added tests and benchmarks.

yan-alex committed:
Working templated function for current validity tests

yan-alex committed:
WIP, tests failing. SetInvalidRange receives Vector instead of ValidityMask

yan-alex committed:
Fix normal roaring tests

yan-alex committed:
make format-fix

yan-alex committed:
D_ASSERT(result_data);

yan-alex committed:
Merge remote-tracking branch 'origin/main' into #6429-roaring-booleans

yan-alex committed:
Revert "Fix normal roaring tests"

This reverts commit 9ea9e414fbe83c16ef3c0bf6b9784022bb772de4.

# Conflicts:
#	src/storage/compression/roaring/common.cpp

yan-alex committed:
Undo templating, use RoaringScanBoolean

yan-alex committed:
Update numeric stats, roaring bool working

yan-alex committed:
tests

yan-alex committed:
Handle null validity mask with memset 1

yan-alex committed:
Use handrolled bitpacking code to increase underlying boolean run lengths throughout null values, for RLE

yan-alex committed:
Fix handrolled bitpacking checking previous bool bug. It triggers an D_ASSERT on roaring_bool_array_simple_w_null.test

yan-alex committed:
More tests, switch to BitPackingPrimitives because of array test error

yan-alex committed:
Fix test

yan-alex committed:
passing, but array test fails w handroll

yan-alex committed:
FIX: Forgot to bitpack booleans on analyze step

yan-alex committed:
Check validity mask to ensure null value when bitpacking

yan-alex committed:
Coherent bool tests

yan-alex committed:
restructure

yan-alex committed:
minor test fix

yan-alex committed:
Merge branch 'main' into #6429-roaring-booleans

yan-alex committed:
cleanup

Tishj commented on src/include/duckdb/common/bitpacking.hpp:
```diff
@@ -112,6 +112,34 @@ class BitpackingPrimitives {
+	inline static void BitPackBooleans(data_ptr_t dst, const data_ptr_t src, const idx_t count,
+	                                   const ValidityMask &validity_mask) {
+		bool is_last_bit_true = false; // If first value is null, write false, as it's probably the most common value
+		for (idx_t i = 0; i < count; i++) {
```
I think we can predicate a lot from this loop (make it branchless)
updating `is_last_bit_true` can be outside of the if/else tree for starters
Maybe have a look at some of the bitpacking internals to see how they handle the bitpacking logic, I imagine that's very well optimized

Tishj commented on src/storage/compression/roaring/common.cpp:
```diff
@@ -198,9 +198,16 @@ unique_ptr<CompressionState> RoaringInitCompression(ColumnDataCheckpointData &ch
+	if (TYPE == PhysicalType::BOOL) {
+		auto data_ptr = scan_vector.GetData();
+		for (idx_t i = 0; i < count; i++) {
+			state.current_segment->stats.statistics.UpdateNumericStats<bool>(data_ptr[i]);
```
I wonder if we can somehow push this into the bitpacking step, since we're already looping over the vector in its entirety there as well

yan-alex committed:
Update statistics in the bitpacking loop

yan-alex committed:
Simplify code

yan-alex commented on src/storage/compression/roaring/common.cpp:
```diff
@@ -259,16 +284,37 @@ unique_ptr<CompressedSegmentState> RoaringInitSegment(ColumnSegment &segment, bl
+	case PhysicalType::BOOL: {
+		analyze = roaring::RoaringAnalyze<PhysicalType::BOOL>;
+		compress = roaring::RoaringCompress<PhysicalType::BOOL>;
+		scan = roaring::RoaringScanBoolean;
```
True... But weird that it passed all the tests... Maybe the current tests aren't covering enough then

yan-alex committed:
Added bool version for FetchRow and ScanPartial, moved BitPackBooleans out of bitpacking.hpp

yan-alex commented on src/include/duckdb/common/bitpacking.hpp:
```diff
@@ -112,6 +112,34 @@ class BitpackingPrimitives {
 
+	inline static void BitPackBooleans(data_ptr_t dst, const data_ptr_t src, const idx_t count,
+	                                   const ValidityMask &validity_mask) {
+		bool is_last_bit_true = false; // If first value is null, write false, as it's probably the most common value
```
> ```c++
> 		// Fast path: stats have already been set, and there's no NULLs
> 		for (idx_t r = chunk_start; r < chunk_end; r++) {
> 			const auto &val = ptr[r];
> 			state.byte |= val << state.byte_pos;
> 			if (++state.byte_pos == 8) {
> 				temp_writer.Write(state.byte);
> 				state.byte = 0;
> 				state.byte_pos = 0;
> 			}
> 		}
> ```

For `state.byte |= val << state.byte_pos;` to work, I would have to first set all the data in dst to zero. Is it worth it timewise to do so?

yan-alex committed:
Optimize bitpacking

yan-alex committed:
More bitpacking optimizations

yan-alex commented on benchmark/micro/compression/roaring/roaring_bool_array_read.benchmark:
```diff
@@ -0,0 +1,24 @@
+
+name Roaring Scan Array Container
+group roaring
+storage persistent v1.2.0
```
Should I change it to a different version? If so, which one?

Mytherin reviewed (commented):
Thanks! Looks good - some comments from my side

Tishj commented on test/sql/storage/compression/roaring/roaring_bool_first_is_null.test:
```diff
@@ -0,0 +1,67 @@
+CREATE TABLE test (a BOOL);
+
+statement ok
+INSERT INTO test VALUES
```
```
query I nosort expected_res
FROM uncompressed_booleans
----

query I nosort expected_res
FROM roaring_booleans
----

reset label expected_res
```

Could also be used, in case that's faster than doing a EXCEPT_ALL query

And also I suggest adding a `select setseed(42);` to the file, to make failures reproducable

yan-alex replied:
Added a test for this. I would like to keep this test though if that's ok, as it explicitly starts with a null

yan-alex replied:
> Could also be used, in case that's faster than doing a EXCEPT_ALL query

I tried it, it was actually slower.

Added seed though.

yan-alex committed:
Remove unnecessary bitwise OR,  revert appender changes

Reverting appender changes, as to have a smaller change footprint, pass all data to it as validity masks.

yan-alex committed:
Restrict roaring booleans to v1.5.0 or higher. Fail on compile time when using Analyze with unsupported Physical Type

yan-alex committed:
refactor: streamline boolean bitpacking functions

Consolidate the separate `BitPackBooleansNoNulls` and
`BitPackBooleansWithNulls` functions into a single templated function
`BitPackBooleans`. This unifies handling boolean bitpacking with and
without nulls, improving maintainability. The function now uses templates
to update statistics conditionally, avoiding repeated code, maintaining mostly the same performance

yan-alex committed:
Add test with generated data

yan-alex committed:
Merge branch 'main' into #6429-roaring-booleans

yan-alex committed:
test changes

yan-alex committed:
Static assert for compress as well

Tishj reviewed (changes requested):
Thanks for the fixes, looks great!
Just one last request for a test verifying that with storage version v1.4.0 using force_compression='Roaring' on a BOOLEAN column results in Uncompressed, to make sure the check we put in works

yan-alex committed:
Added test for version 1.4

yan-alex committed:
Merge remote-tracking branch 'upstream/main' into #6429-roaring-booleans

Tishj reviewed (approved):
Thanks!
